src/DIRAC/Interfaces/API/Dirac.py

```diff
@@ -324,33 +325,41 @@ def submitJob(self, job, mode="wms"):
             else:
                 self.log.verbose("Job is a JDL string")
                 jdlAsString = job
-            jobDescriptionObject = None
-        else:  # we assume it is of type "DIRAC.Interfaces.API.Job.Job"
-            try:
-                formulationErrors = job.errorDict
-            except AttributeError as x:
-                self.log.verbose("Could not obtain job errors:%s" % (x))
-                formulationErrors = {}
-
-            if formulationErrors:
-                for method, errorList in formulationErrors.items():  # can be an iterator
-                    self.log.error(">>>> Error in %s() <<<<\n%s" % (method, "\n".join(errorList)))
-                return S_ERROR(formulationErrors)
-
-            # Run any VO specific checks if desired prior to submission, this may or may not be overridden
-            # in a derived class for example
-            try:
-                result = self.preSubmissionChecks(job, mode)
-                if not result["OK"]:
-                    self.log.error('Pre-submission checks failed for job with message: "%s"' % (result["Message"]))
-                    return result
-            except Exception as x:
-                msg = 'Error in VO specific function preSubmissionChecks: "%s"' % (x)
-                self.log.error(msg)
-                return S_ERROR(msg)
 
-            jobDescriptionObject = io.StringIO(job._toXML())  # pylint: disable=protected-access
-            jdlAsString = job._toJDL(jobDescriptionObject=jobDescriptionObject)  # pylint: disable=protected-access
+            result = self.parseJobFromJDL(jdlAsString)
+            if not result["OK"]:
+                self.log.error("Error while parsing the jdl string", result["Message"])
+                return result
+            job = result["Value"]
+
+        if not isinstance(job, Job):
+            return self._reportError("Job must be a str or a job object")
+
+        try:
+            formulationErrors = job.errorDict
+        except AttributeError as x:
+            self.log.verbose("Could not obtain job errors:%s" % (x))
+            formulationErrors = {}
+
+        if formulationErrors:
+            for method, errorList in formulationErrors.items():  # can be an iterator
+                self.log.error(">>>> Error in %s() <<<<\n%s" % (method, "\n".join(errorList)))
+            return S_ERROR(formulationErrors)
+
+        # Run any VO specific checks if desired prior to submission, this may or may not be overridden
+        # in a derived class for example
+        try:
+            result = self.preSubmissionChecks(job, mode)
+            if not result["OK"]:
+                self.log.error('Pre-submission checks failed for job with message: "%s"' % (result["Message"]))
```

Comment:
Normally we would do the opposite, and anyway you can use f-strings:
```suggestion
                self.log.error(f"Pre-submission checks failed for job with message: {result['Message']}'")
```